common/src/gas_provider/internal.rs

```diff
@@ -368,19 +374,25 @@ where
         }
 
         // Detect inner from `reserve`.
-        let new_node = if reserve {
-            let id = Self::get_external(key)?;
-
-            GasNode::ReservedLocal { id, value: amount }
-        } else {
-            node.increase_spec_refs();
-
-            GasNode::SpecifiedLocal {
-                value: amount,
-                lock: Zero::zero(),
-                parent: node_id,
-                refs: Default::default(),
-                consumed: false,
+        let new_node = match kind {
```

Comment:
I'd move this lines under the following ones: try to reduce value, create node afterward

It's minor optimisation if insufficient balance queried, no `Self::get_external` will be called.